### PR TITLE
Modify the instructions emited for PREFETCH on arm64

### PR DESCRIPTION
**Pull request by AGSaidi**
__builtin_prefetch(...., 1) prefetches into the L2 cache on x86 while the same
emits a pldl3keep instruction on arm64 which doesn't seem to be close enough.

Testing on a Graviton3, and M1 system with memtablerep_bench fillrandom and
skiplist througpuh increased as follows adjusting the 1 to 2 or 3:
```
           1 -> 2     1 -> 3
----------------------------
Graviton3   +10%        +15%
M1          +10%        +10%
```

Given that prefetching into the L1 cache seems to help, I chose that conversion

### CODE REVIEW

**Comment by siying** [comment]
The use in inlineskiplist isn't heavy and the chance of prefetch useful is high. I think we probably should change it to 3. I'm not sure about the bloom filter and ribbon filter cases. CC @pdillinger

**Comment by AGSaidi** [comment]
filterbench also seems to show benefit with this change on a Graviton3 system experimenting with Ribbon128

**Comment by pdillinger** [comment]
If I understand correction, increasing the locality value is not intended to make the prefetched value more immediately available, but more likely to stay around for subsequent accesses.

> A value of zero means that the data has no temporal locality, so it need not be left in the cache after the access. A value of three means that the data has a high degree of temporal locality and should be left in all levels of cache possible. Values of one and two mean, respectively, a low or moderate degree of temporal locality. The default is three.

https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html

Am I misreading this? We typically use PREFETCH with locality=1 to mean, "I'm almost certainly going to use this value very soon, but that's not really predictive (positively or negatively) of subsequent uses."

But regardless, if increasing the locality value does in practice make the prefetched value more immediately available, then yes I think we should do that, for platforms on which that is true. Are you interested in investigating @mdcallag?

**Comment by AGSaidi** [comment]
On 64b Arm processors locality=1 turns into a pldl3keep, 2=>pldl2keep, 3=>pldl1keep. On Arm these are supposed to target the L3, L2, and L1 cache respectively.  A locality of 0 is prefetched as a pld1strm meaning "Streaming or non-temporal prefetch, for data that is used only once.". Of course these are considered hints, so in reality the implementation can ignore them.

On x86 these turn into a PREFETCHNTA/T0/T1/T2. The exact meaning for T0/1/2 is vague but it seems like T0 means L1, and T1/T2 means L2 or higher. 

While the intent for the built in might have been to communicate some level of re-use, in practice these instructions are really placing data into a particular level of the cache not speaking to the general temporal patterns beyond not-temporal.

**Comment by siying** [comment]
memtablerep_bench has big variation. I ran many times but still can't tell whether there is an improvement 1->3 on x86. I guess it's hard to say it can benefit x86.

File: port/port_posix.h [hunk 1]
```diff
@@ -202,7 +202,14 @@ extern void *cacheline_aligned_alloc(size_t size);
 
 extern void cacheline_aligned_free(void *memblock);
 
+#if defined(__aarch64__)
+//  __builtin_prefetch(..., 1) turns into a prefetch into prfm pldl3keep. On
+//  arm64 get these into the L1 and L2
+#define PREFETCH(addr, rw, locality) \
+  __builtin_prefetch(addr, rw, locality == 1 ? 3 : locality)
```
Comment (siying) [inline port/port_posix.h:209]: The mapping still feels odd to me:
0->0, 1->3, 2->2, 3->3.

It's not monotonic increasing.

Comment (AGSaidi) [reply port/port_posix.h:209]: 0 is special, it means non-temporal.  I could find any uses in the code of 2, so one way would be to just write:
```locality != 0 ? 3 : 0``` which would mean 0->0; 1,2,3->3

Comment (siying) [reply port/port_posix.h:209]: > 

I think it feels better. If we do 1->2, what the performance would be?

**Comment by AGSaidi** [comment]
> It means in x86, 1 is probably the fastest in terms of latency, while in ARM, it does almost nothing?

I believe on x86 1 means L2, 2 means L1 or L2 and 3 means L1 but i haven't found something I'd consider definitive. We could try and figure it out looking at performance counters, but I don't think it's that useful.

I did try running the following `./memtablerep_bench -benchmarks fillrandom -memtablerep skiplist  -num_operations 1000000`  with a change that increased 1 to 2, or increased 1 to 3 on a otherwise idle icelake-based system.

Averaging 20 runs there is still more noise than signal with changing 1->3. 

```
run number      1->2       1->3
-------------------------------
1               -0.4%      -0.3%
2               1.0%       2.0%
3               0.0%       0.6%
```

File: port/port_posix.h (hunk 1 above)
Comment (AGSaidi) [reply port/port_posix.h:209]: We're still leaving a fair bit (5%) on the table...
```
           1 -> 2     1 -> 3
----------------------------
Graviton3   +10%        +15%
M1          +10%        +10%
```

To be actually used the data is going to end up in the L1 cache, so as long as we're not over prefetching (e.g. prefetching data we have a low chance of utilizing) and the use is going to be "soon" then the line is going to be making it's way into the L1 shortly and we might as well help it along.

Comment (siying) [reply port/port_posix.h:209]: OK. Either 1->2, or 1,2->3 sounds good to me.

Comment (AGSaidi) [reply port/port_posix.h]: ack, updated.

**Review by siying — APPROVED** [review]

**Comment by siying** [comment]
@AGSaidi are you able to sign CLA? https://code.facebook.com/cla

**Comment by AGSaidi** [comment]
> @AGSaidi are you able to sign CLA? https://code.facebook.com/cla

Working on getting added to it

**Comment by AGSaidi** [comment]
I’ve been added to Amazon’s existing signed cla. Is there a way to rerun the check?

**Comment by AGSaidi** [comment]
Looks like the CLA bot picked it up eventually

**Pull request closed without merging**